sw/host/tests/chip/sram_ctrl/src/sram_ctrl_lc_escalation.rs

```diff
@@ -0,0 +1,143 @@
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "600s")]
+    timeout: Duration,
+
+    /// Path to the ELF file being tested on the device.
+    #[arg(long)]
+    firmware_elf: PathBuf,
+}
+
+/// Addresses of symbols in the ELF file that we will access.
+///
+/// Note that the `main` and `ret` addresses point to variables holding the
+/// main and retention SRAM addresses and not those addresses directly.
+#[derive(Debug, Clone, Copy)]
+struct Addresses {
+    test_phase: u32,
+    main: u32,
+    ret: u32,
+}
+
+fn main() -> Result<()> {
+    let opts = Opts::parse();
+    opts.init.init_logging();
+
+    let elf_file = fs::read(&opts.firmware_elf).context("failed to read ELF")?;
+    let object = object::File::parse(elf_file.as_ref()).context("failed to parse ELF")?;
+
+    let addresses = Addresses {
+        test_phase: test_utils::object::symbol_addr(&object, "test_phase")?,
+        main: test_utils::object::symbol_addr(&object, "sram_buffer_addr_main")?,
+        ret: test_utils::object::symbol_addr(&object, "sram_buffer_addr_ret")?,
+    };
+
+    let transport = opts.init.init_target()?;
+    let uart_console = transport.uart("console")?;
+
+    execute_test!(lc_escalation, &opts, &transport, &*uart_console, addresses);
+
+    Ok(())
+}
+
+/// Send and receive data with a device's UART.
+fn lc_escalation(
+    opts: &Opts,
+    transport: &TransportWrapper,
+    console: &dyn Uart,
+    addresses: Addresses,
+) -> Result<()> {
+    UartConsole::wait_for(console, r"waiting for commands", opts.timeout)?;
+
+    // Get the addresses of main and retention SRAM that we can access.
+    let main_addr = MemRead32Req::execute(console, addresses.main)?;
+    let ret_addr = MemRead32Req::execute(console, addresses.ret)?;
+
+    // Enable JTAG debugging.
+    transport.pin_strapping("PINMUX_TAP_RISCV")?.apply()?;
+    let mut jtag = opts
+        .init
+        .jtag_params
+        .create(transport)?
+        .connect(JtagTap::RiscvTap)?;
+
+    // Test that we can write to and read back the SRAMs over the debugger.
+    write_read(&mut *jtag, main_addr, 0xaaaa_aaaa)?;
+    write_read(&mut *jtag, ret_addr, 0xbbbb_bbbb)?;
+
+    // Let the device side of the test escalate an alert and lock up.
+    MemWriteReq::execute(console, addresses.test_phase, &[1])?;
+
+    log::info!("-------------------------------");
+    log::info!("EXPECTING JTAG ACCESSES TO FAIL");
+    log::info!("-------------------------------");
+
+    // Check that we can no longer access the SRAMs after escalation.
+    // These operations should time out.
+    ensure!(
+        write_read(&mut *jtag, main_addr, 0xcccc_cccc).is_err(),
+        "expected main SRAM access to fail"
+    );
+    ensure!(
+        write_read(&mut *jtag, ret_addr, 0xdddd_dddd).is_err(),
+        "expected retention SRAM access to fail"
+    );
```

Comment:
Is it worth checking using register that there was an escalation alert?

Reply:
Good idea

Do you mean trying to read the alert handler's registers to see whether the alert was triggered, or just trying to read any generic register or CSR to make sure the debugger still works?

Reply:
Read the alert handler's register

Reply:
This turned out not to be as easy as I expected, do you mind if I do it in a follow-up PR?